framework/reporter.go

```diff
@@ -0,0 +1,278 @@
+/*
+ Copyright 2024 The OSCAL Compass Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package framework
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/defenseunicorns/go-oscal/src/pkg/uuid"
+	oscalTypes "github.com/defenseunicorns/go-oscal/src/types/oscal-1-1-2"
+	"github.com/hashicorp/go-hclog"
+	"github.com/oscal-compass/compliance-to-policy-go/v2/framework/config"
+	"github.com/oscal-compass/compliance-to-policy-go/v2/policy"
+	"github.com/oscal-compass/oscal-sdk-go/extensions"
+	"github.com/oscal-compass/oscal-sdk-go/models"
+	"github.com/oscal-compass/oscal-sdk-go/rules"
+	"github.com/oscal-compass/oscal-sdk-go/settings"
+)
+
+const (
+	defaultVersion = "0.1.0"
+	defaultTitle   = "Automated Assessment Results"
+)
+
+type Reporter struct {
+	log        hclog.Logger
+	rulesStore rules.Store
+}
+
+func NewReporter(cfg *config.C2PConfig) (*Reporter, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	rulesStore, _, err := config.ResolveOptions(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Reporter{
+		log:        cfg.Logger,
+		rulesStore: rulesStore,
+	}, nil
+}
+
+type generateOpts struct {
+	title string
+}
+
+func (g *generateOpts) defaults() {
+	g.title = models.DefaultRequiredString
+}
+
+// GenerateOption defines optional arguments to tune the behavior of GenerateAssessmentResults
+type GenerateOption func(opts *generateOpts)
+
+// WithTitle is a GenerateOptions that sets the AssessmentResults title in the metadata
+func WithTitle(title string) GenerateOption {
+	return func(opts *generateOpts) {
+		opts.title = title
+	}
+}
+
+// Generate OSCAL Findings for all non-passing controls in the OSCAL Observation
+func (r *Reporter) generateFindings(observation oscalTypes.Observation, ruleSet extensions.RuleSet, implementationSettings settings.ImplementationSettings) ([]oscalTypes.Finding, error) {
+
+	findings := make([]oscalTypes.Finding, 0)
+
+	applicableControls, err := implementationSettings.ApplicableControls(ruleSet.Rule.ID)
+	if err != nil {
+		return findings, err
+	}
+
+	for _, controlId := range applicableControls {
+
+		targetId := fmt.Sprintf("%s_smt", controlId)
```

Comment:
Nit: To make the results JSON a bit easier to read, I would suggest indexing the findings as they are created so you don't end up creating multiple finding for the same control.